src/snowflake/snowpark/_internal/compiler/utils.py

```diff
@@ -323,3 +327,75 @@ def get_snowflake_plan_queries(
 def is_active_transaction(session):
     """Check is the session has an active transaction."""
     return session._run_query("SELECT CURRENT_TRANSACTION()")[0][0] is not None
+
+
+def plot_plan_if_enabled(root: TreeNode, filename: str) -> None:
+    """A helper function to plot the query plan tree using graphviz useful for debugging.
+    It plots the plan if the environment variable ENABLE_SNOWFLAKE_OPTIMIZATION_PLAN_PLOTTING
+    is set to true.
+
+    The plots are saved in the temp directory of the system which is obtained using
+    https://docs.python.org/3/library/tempfile.html#tempfile.gettempdir. Setting env variable
+    TMPDIR to your desired location is recommended. Within the temp directory, the plots are
+    saved in the directory `snowpark_query_plan_plots` with the given `filename`.
+
+    Args:
+        root: root TreeNode of the plan to plot.
+        filename: name of the file to save the image plot.
+    """
+    import os
+
+    import graphviz  # pyright: ignore[reportMissingImports]
+
+    if (
+        os.environ.get("ENABLE_SNOWFLAKE_OPTIMIZATION_PLAN_PLOTTING", "false").lower()
+        != "true"
+    ):
+        return
+
+    def get_stat(node: TreeNode):
+        def get_name(node: Optional[LogicalPlan]) -> str:
+            if node is None:
+                return "EMPTY_SOURCE_PLAN"
+            addr = hex(id(node))
+            name = str(type(node)).split(".")[-1].split("'")[0]
+            return f"{name}({addr})"
+
+        name = get_name(node)
+        if isinstance(node, SnowflakePlan):
+            name = f"{name} :: ({get_name(node.source_plan)})"
+        elif isinstance(node, SelectSnowflakePlan):
+            name = f"{name} :: ({get_name(node.snowflake_plan.source_plan)})"
+        elif isinstance(node, SetStatement):
+            name = f"{name} :: ({node.set_operands[1].operator})"
+
+        score = get_complexity_score(node.cumulative_node_complexity)
+        sql_text = (
+            node.queries[-1].sql if isinstance(node, SnowflakePlan) else node.sql_query
+        )
+        sql_size = len(sql_text)
+        sql_preview = sql_text[:50]
+
+        return f"{name=}\n{score=}, {sql_size=}\n{sql_preview=}"
+
+    g = graphviz.Graph(format="png")
+
+    curr_level = [root]
+    edges = set()  # add edges to set for de-duplication
+    while curr_level:
+        next_level = []
+        for node in curr_level:
+            node_id = hex(id(node))
+            g.node(node_id, get_stat(node))
+            for child in node.children_plan_nodes:
```

Comment:
in that case, does it mean that for snowflake plan node, we will not plot the source plan node? but the child direclty?

Reply:
yeah. the source plan node will not show up as a node but it will show up as a child in the name.
For example see `Aggregate` is a `LogicalPlan` node associated with the `SelectSnowflakePlan`. In reality, we have `SelectSnowflakePlan` -> `SnowflakePlan` -> `Aggregate`

<img width="671" alt="Screenshot 2024-09-25 at 4 56 12 PM" src="https://github.com/user-attachments/assets/dde5500f-25f1-49ba-88cb-8162a5bfbf0e">